src/internal/operators/share.ts

```diff
@@ -84,88 +101,151 @@ export function share<T>(options: ShareConfig<T>): MonoTypeOperatorFunction<T>;
-        error: (err: any) => {
-          hasErrored = true;
-          // We need to capture the subject before
-          // we reset (if we need to reset).
-          const dest = subject!;
-          if (resetOnError) {
-            reset();
-          }
-          dest.error(err);
-        },
-        complete: () => {
-          hasCompleted = true;
-          const dest = subject!;
-          // We need to capture the subject before
-          // we reset (if we need to reset).
-          if (resetOnComplete) {
-            reset();
-          }
-          dest.complete();
-        },
+
+      // Create the subject if we don't have one yet. Grab a local reference to
+      // it as well, which avoids non-null assertations when using it and, if we
+      // connect to it now, then error/complete need a reference after it was
+      // reset.
+      const dest = (subject = subject ?? connector());
+
+      // Add the teardown directly to the subscriber - instead of returning it -
+      // so that the handling of the subscriber's unsubscription will be wired
+      // up _before_ the subscription to the source occurs. This is done so that
+      // the assignment to the source connection's `closed` property will be seen
+      // by synchronous firehose sources.
+      subscriber.add(() => {
+        refCount--;
+
+        // If we're resetting on refCount === 0, and it's 0, we only want to do
+        // that on "unsubscribe", really. Resetting on error or completion is a different
+        // configuration.
+        if (refCount === 0 && !hasErrored && !hasCompleted) {
+          resetConnection = handleReset(resetAndUnsubscribe, resetOnRefCountZero);
+        }
       });
-      from(source).subscribe(connection);
-    }
-  });
+
+      // The following line adds the subscription to the subscriber passed.
+      // Basically, `subscriber === dest.subscribe(subscriber)` is `true`.
+      dest.subscribe(subscriber);
+
+      if (!connection) {
+        // We need to create a subscriber here - rather than pass an observer and
+        // assign the returned subscription to connection - because it's possible
+        // for reentrant subscriptions to the shared observable to occur and in
+        // those situations we want connection to be already-assigned so that we
+        // don't create another connection to the source.
+        connection = new SafeSubscriber({
+          next: (value) => dest.next(value),
+          error: (err) => {
+            hasErrored = true;
+            cancelReset();
+            resetConnection = handleReset(reset, resetOnError, err);
+            dest.error(err);
+          },
+          complete: () => {
+            hasCompleted = true;
+            cancelReset();
+            resetConnection = handleReset(reset, resetOnComplete);
+            dest.complete();
+          },
+        });
+        from(source).subscribe(connection);
+      }
+    })(wrapperSource);
```

Comment:
is it worth introducing a helper function to make this more clear?

something like:
```ts
export function share<T>(options: ShareConfig<T> = {}): MonoTypeOperatorFunction<T> {
  // init operator

  return prepareAndOperate(() => {
    // init pipeline

    return (source, subscriber) => {
      // init subscription
    };
  });
}
```

Reply:
I don't think that a helper function is warranted, as four of these operators are going to be removed in version 8. After that, AFAICT, `share` will be the only operator that returns an operator function that's not referentially transperant.